storage/aptosdb/src/state_merkle_db.rs

```diff
@@ -286,88 +291,160 @@ impl StateMerkleDb {
         Ok(None)
     }
 
-    /// Merklize the results generated by `value_state_sets` to `batch` and return the result root
-    /// hashes for each write set.
+    fn create_jmt_commit_batch_for_shard(
+        &self,
+        shard_id: Option<u8>,
+        tree_update_batch: &TreeUpdateBatch<StateKey>,
+        previous_epoch_ending_version: Option<Version>,
+    ) -> Result<SchemaBatch> {
+        let _timer = OTHER_TIMERS_SECONDS
+            .with_label_values(&["create_jmt_commit_batch_for_shard"])
+            .start_timer();
+
+        let batch = SchemaBatch::new();
+
+        tree_update_batch
+            .node_batch
+            .iter()
+            .flatten()
+            .collect::<Vec<_>>()
+            .par_iter()
+            .with_min_len(128)
+            .try_for_each(|(node_key, node)| {
+                ensure!(node_key.get_shard_id() == shard_id);
+                batch.put::<JellyfishMerkleNodeSchema>(node_key, node)
+            })?;
+
+        tree_update_batch
+            .stale_node_index_batch
+            .iter()
+            .flatten()
+            .collect::<Vec<_>>()
+            .par_iter()
+            .with_min_len(128)
+            .try_for_each(|row| {
+                ensure!(row.node_key.get_shard_id() == shard_id);
+                if previous_epoch_ending_version.is_some()
+                    && row.node_key.version() <= previous_epoch_ending_version.unwrap()
+                {
+                    batch.put::<StaleNodeIndexCrossEpochSchema>(row, &())
+                } else {
+                    // These are processed by the state merkle pruner.
+                    batch.put::<StaleNodeIndexSchema>(row, &())
+                }
+            })?;
+
+        Ok(batch)
+    }
+
+    #[cfg(test)]
     pub fn merklize_value_set(
         &self,
         value_set: Vec<(HashValue, Option<&(HashValue, StateKey)>)>,
+        version: Version,
+        base_version: Option<Version>,
+        previous_epoch_ending_version: Option<Version>,
+    ) -> Result<(SchemaBatch, Vec<SchemaBatch>, HashValue)> {
+        let mut sharded_value_set: Vec<Vec<(HashValue, Option<&(HashValue, StateKey)>)>> =
+            Vec::new();
+        sharded_value_set.resize(NUM_STATE_SHARDS, Default::default());
+        value_set.into_iter().for_each(|(k, v)| {
+            sharded_value_set[k.nibble(0) as usize].push((k, v));
```

Comment:
this implies each index of value_set corresponds to a shard_id and value_set here contain 16 shards' value respectively?  if so, better to mention this in the function comment to prevent people use it incorrectly.

Reply:
No. The input contains all updates from all shards, and the logic here is to group them into sharded_value_set by shard_id (k.nibble(0) here).

Comment added.

Reply:
have a function `get_shard_id` for this to make it easier to understand